add local workflow page (#322)
hubverse-org/hubDocs: docs/source/developer/dashboard-local.md

    dashboard["dashboard repository"]
    ptc/data>ptc data]
    predevals/data>evals data]
    subgraph pre-compute toolchain
      hub-dashboard-predtimechart[\hub-dashboard-predtimechart/]
      hubPredEvalsData[\hubPredEvalsData/]
      hub-dash-site-builder[\hub-dash-site-builder/]
    end
    subgraph site
        index.html([index.html])
        forecast.html([forecast.html])
        eval.html([eval.html])
    end
    dashboard -->|site-config.yml| hub-dash-site-builder
    hub-dash-site-builder --> site
    hub -->|time-series data|hub-dashboard-predtimechart
    dashboard -->|predtimechart-config.yml|hub-dashboard-predtimechart
    hub-dashboard-predtimechart --> ptc/data
    hub -->|oracle data|hubPredEvalsData
    dashboard -->|predevals-config.yml|hubPredEvalsData
    hubPredEvalsData --> predevals/data
    predevals/data -.-> eval.html
    ptc/data -.-> forecast.html
```

## Purpose

This page will demonstrate that workflow using a practical example of building [the flu metrocast dashboard](https://reichlab.io/metrocast-dashboard) locally.
The important takeaway from this exercise is that it exists
**independent of GitHub workflows**. We are not concerned

> about where data are stored or when they are updated.

Reviewer: This is confusing me a little as below we discuss locations of data? Could you clarify what you mean about not being concerned about where data are stored?

Reviewer: Reading on I think you mean not being concerned about **configuring the source and automatic import** of the required data? Instead, locally we can import these manually to the expected locations?

Author: That's a very good point, which means that I need to reframe a bit (this is clearly a case of my over-connected mental map affecting my writing).

My aim is to try to highlight the specific build aspects of the workflow without being dependent on a specific vendor for continuous integration. That is, how do you get a website if you have access to the basic tools.

Once you know the foundations of building this website from sources, then you can start asking questions about how you want to deploy that website and how you make the deploy process usable by others.

If I were to walk through the GitHub workflows first, it would be _easily_ overwhelming because they involve intricacies of operations that would just add to the cognitive load:

1. When should the data and site be built? (push, schedule, pull request, manual dispatch)
2. Where should the data be stored---deploying the data and the website publicly (this involves orphan branches)
3. separating build and deployment steps for security reasons
4. deploying the workflows to repositories we do not own (re-usable workflows)

## Rationale

Building a dashboard website requires three tools:

1. one to build the static website (<https://github.com/hubverse-org/hub-dash-site-builder>)
1. one to build the predtimechart data and options (<https://github.com/hubverse-org/hub-dashboard-predtimechart>)
1. one to build the predevals data and options (<https://github.com/hubverse-org/hubPredEvalsData-docker>)

These take data from two sources:

1. the dashboard repository (<https://github.com/reichlab/metrocast-dashboard>)
2. the hub repository (<https://github.com/reichlab/flu-metrocast>)

The GitHub workflows bind all of these together, but these end
up being complex because they deal with **issues of deployment**
that is, where the data are stored and when they are updated.
These issues are a separate layer of complexity that can be
addressed once you understand the underlying build process.

## Required Software

* BASH --- This is how we will be running everything.

> * [gh](#dashboard-tool-gh)     --- needed for interacting with the GitHub API

Reviewer: Just to clarify, this is referring to a command line tool, not the `gh` R package correct?

* [yq](#dashboard-tool-yq)     --- needed to get the hub information from the site config

> * [uv](https://docs.astral.sh/uv/)     --- python client, runs hub-dashboard-predtimechart

Reviewer: Might be useful to note that the role of `uv` is dependency management?

Author: Good point! In fact, it's mostly to try to avoid the whole python paths problem if I were to assume that everyone has the right version of python installed.

* git    --- needed for cloning repositories
* tree   --- displays a graphical structure of the folders
* [docker](#dashboard-tool-docker) --- runs [hub-dash-site-builder](https://github.com/hubverse-org/hub-dash-site-builder) and [hubPredEvalsData-docker](https://github.com/hubverse-org/hubPredEvalsData-docker).

## Expected Results

After this demo, we will end up with a temporary directory that
has the following structure

```
├── dashboard
│   ├── _site #............ generated by hub-dash-site-builder
│   │   └── resources
│   │       ├── css
│   │       ├── evals     # local copies
│   │       └── forecasts #
│   ├── data
│   │   ├── evals # ....... generated by hubPredEvalsData-docker
│   │   │   └── scores
│   │   └── ptc   # ....... generated by hub-dashboard-predtimechart
│   │       ├── forecasts
│   │       └── targets
│   └── pages
└── hub
    ├── auxiliary-data
    ├── hub-config
    ├── model-metadata
    ├── model-output
    ├── raw-data
    ├── src
    └── target-data
```

|folder|contents|generated by|
|:-----|:-------|-----------:|

> | `dashboard/` | source for the hub dashboard | humans |

Reviewer: what does source for the hub mean exactly? and how do humans create it?

Author: Good question! When you use the hub-dashboard-template to create a new dashboard repository, this is the source for the hub dashboard. As long as you have a repository with a `pages/index.qmd` file and a `site-config.yml` file, then that serves as the bare minimum source for a website.

I should probably expand this in text.

  this workflow. You can find a copy of the script at [zkamvar's gist,
  `2025-04-24-dashboard-demo.sh`](https://gist.github.com/zkamvar/94ee15b486821fa8ef99b81a4098a39c).

(dashboard-local-setup)=
### Setup (source material)

The setup involves **preparing our source materials**, that is our **dashboard
repository** and our **modeling hub**. As shown in [the workflow
diagram](#local-workflow), you can see that the tools need contents from both
repositories in order to work.

1. create a temporary directory and save it in `$tmp`; define the dashboard
   repository
   ```bash
   tmp=$(mktemp -d)
   dash_repo="reichlab/metrocast-dashboard"
   ```
2. clone the dashboard into `$tmp/dashboard`
   ```bash
   dash="$tmp/dashboard"
   git clone "https://github.com/${dash_repo}.git" "$dash"
   ```
3. get the hub name from the dashboard and clone it into `$tmp/hub`
   ```bash
   hub="$tmp/hub"
   repo=$(yq '.hub' "$dash/site-config.yml")
   git clone "https://github.com/${repo}.git" "$hub"
   ```

After these steps, the structure of our temporary folder looks
like this where it contains both the dashboard and the hub
together:

```bash
tree -d -L 4 -I 'epi*|site_libs' $tmp
```
```

> ├── dashboard

Reviewer: I know this focuses on showing directories but I think it would be more instructive to show the files created too?

Reviewer: If showing files makes it look too busy it would be useful to clarify that we are looking at directories and the files generated are ommited

Author: I'm going to go with the second option. The files in the hub would be way too distracting.

│   └── pages
└── hub
   ├── auxiliary-data
   ├── hub-config
   ├── model-metadata
   ├── model-output
   ├── raw-data
   ├── src
   └── target-data
```

:::{admonition} Forward Thinking

> The state of our directory is for us to have both the hub and the dashboard.

Reviewer: This sentence sounds a little awkward?

Author: This was probably an end-of-day-and-my-eyes-are-melting kind of sentence. 

```suggestion
We currently have both the hub and the dashboard sources in our temporary working folder.
```

At the moment, both data generation tools assume that the source data can only
be accessed via cloning a hub. However, since we have S3 data available to us,
it should be feasible to fetch the hub as a sparse clone (see the tip in
<https://hubverse.io/quickstart/submit.html>) and fetch the data using S3.

:::

The next two steps generate the data used by the dashboards and can be done in
any order.

(dashboard-local-forecasts)=
### Generate forecasts (python)

[The forecasts
visualization](https://reichlab.io/metrocast-dashboard/forecast.html) is

> provided by [PredTimeChart](https://github.com/reichlab/predtimechart#readme).

Reviewer: Should this repo be part of the hubverse going forward?

Author: That's a @matthewcornell and @nickreich question.

This reads in JSON files that subdivide the data in the format of
`[target]_[location]_[date].json`. For each combination of target, location,
and date, two files are generated: one that records the X and Y values for the
target data and one that records the predictions for each model.

To generate these data, you need three things:

* From the dashboard repository: `predtimechart-config.yml`, the [PredTimeChart configuration file](#dashboard-ptc-config)
* From the hub: [Target timeseries data](#target-time-series)
* From the hub: [Model output data](/user-guide/model-output)

Using the hub and dashboard repositories you cloned in [the setup

> step](#dashboard-local-setup), you can generate the data with these steps. We

Reviewer: ```suggestion
step](#dashboard-local-setup), you can generate the data with the following steps. We
```

need a place to generate the forecasts, so we place them in the `data/ptc`
folder inside the dashboard. The reason why we place them inside the
dashboard and not outside is because the docker container for the website
generation expects the data to be in the same folder.

1. create targets output dir and forecasts outputs dir in `dashboard/data/ptc/`
   ```bash
   mkdir -p $dash/data/ptc/{targets,forecasts}
   ```
2. install
   [hub-dashboard-predtimechart](https://github.com/hubverse-org/hub-dashboard-predtimechart)
   in a virtual environment.
   ```bash
   uv venv --seed "$dash/.venv"
   source "$dash/.venv/bin/activate"

   latest=$(gh api -X GET "repos/hubverse-org/hub-dashboard-predtimechart/releases/latest" --jq ".tag_name")
   uv run pip install --quiet --upgrade pip
   uv run pip install --quiet "git+https://github.com/hubverse-org/hub-dashboard-predtimechart@$latest"
   ```
   This will install a command line application with two functions
   `ptc_generate_target_json_files`---which generates JSON files from target
   data, and `ptc_generate_json_files` which generates both forecast data files

>    and the predtimechart options JSON file.

Reviewer: earlier `predtimechart options JSON` is refered only as an options file. Further down it's called a config file. And so far I don't think there's been clarification of what the file actually contains or is used for? Might be good to add this information and use consistent description of files to avoid confusion.

Author: Ah yeah, consistency in writing is not really my strong suit. 

I think I wanted to avoid getting too deep in the weeds about what the file configures because the important part for this exercise is that the file exists and predtimechart needs it. 

I'll see what I can do to rephrase.

4. generate the target data. This takes three arguments, the hub, the
   configuration file, and finally the output directory.
   ```bash
   ptc_generate_target_json_files \
     $hub \
     $dash/predtimechart-config.yml \
     $dash/data/ptc/targets
   ```

> 5. generate the forecast data and configuration file. This takes four arguments,

Reviewer: ```suggestion
5. generate the forecast data and `predtimechart-options.json` file. This takes four arguments,
```

   the hub, the configuration file, the output options file and the output
   directory for the forecasts.
   ```bash
   ptc_generate_json_files \
     $hub \
     $dash/predtimechart-config.yml \
     $dash/data/ptc/predtimechart-options.json \
     $dash/data/ptc/forecasts
   ```

Once you are done, you will see a folder that looks similar to [the metrocast

> dashboard pdf/data

Reviewer: ```suggestion
dashboard ptc/data
```
?

Author: 🤦🏽 mea culpa

branch](https://github.com/reichlab/metrocast-dashboard/tree/ptc/data).

> 

Reviewer: This is the first time te branches are mentioned. It might be good to add an admonition to describe how the local locations (which do not involve separate branches) relate / differ to data that lives on branches in GitHub dashboards.

Reviewer: Oh I see there's a section that describes this later. Might be good to make a short reference to the fact that we are are linking to data on a separate branch and point to the later section for more details.

Author: This is a good point and I think I'm going to add the admonition at the top where I discuss how we are not going into detail about operations.

(dashboard-local-evals)=
### Generate evals (docker)

[The evaluation
visualization](https://reichlab.io/metrocast-dashboard/eval.html) is

> provided by [PredEvals](https://github.com/reichlab/predevals). This differs

Reviewer: Same comment as above. Should this repo not become formally a part of the hubverse?

Author: Good catch! This _is_ actually part of the hubverse, but I made a copy/paste error

```suggestion
provided by [PredEvals](https://github.com/hubverse-org/predevals). This differs
```

from [PredTimeChart](#dashboard-local-forecasts) in that it expects data in
CSV format that are nested in folders which represents the scores based on
different levels of disaggregation of task IDs. Inside each folder is a file
called `scores.csv`, which contains models in rows and the scores and other
data in columns.

To generate these data, you need three things:

* From the dashboard repository: `predevals-config.yml`, the [PredEvals configuration file](#dashboard-predevals-config)
* From the hub: [Target oracle output data](#target-oracle-output)
* From the hub: [Model output data](/user-guide/model-output)

Using the hub and dashboard repositories you cloned in [the setup
step](#dashboard-local-setup), you can generate the data with these steps.

1. create evals output dir.
   ```bash
   mkdir -p $dash/data/evals
   ```
2. download the docker image (note that the image name needs to be in all lower case)
   ```bash
   docker pull --platform=linux/amd64 ghcr.io/hubverse-org/hubpredevalsdata-docker:main
   ```

> 3. build generate the evaluations.

Reviewer: ```suggestion
3. Generate the evaluations.
```

   ```bash
   docker run --rm -it --platform=linux/amd64 \
     -v "$dash":"/project" \
     -v "$hub":"/hub" \
     ghcr.io/hubverse-org/hubpredevalsdata-docker:main \
     create-predevals-data.R \
       -h "/hub" \
       -c "predevals-config.yml" \
       -d "/hub/target-data/oracle-output.csv" \
       -o "data/evals"
   ```
   There is a lot going on in these lines, so let me break it down:
   1. `docker run --rm -it --platform=linux/amd64` tells docker to
      automatically remove the image when it's done, `-it` means that it's an

>       interactive TTY, and `--platform=linux/amd64` allows it to run on newer

Reviewer: TTY?

Author: Oof, yeah this is one of those things that exist because there used to be physical hardware responsible for the job (similar to why we have "line feed" and "carriage return" ending characters even though we no longer use typewriters)

I might just link this answer on SO: https://stackoverflow.com/a/40026942

>       macOS machines

Reviewer: Thank you for this!!

   2. `-v` means "volume". It's pattern is `-v "/path/on/your/computer":"/path/in/image"`.
      By default, a docker image will run completely independent of your
      computer, but there are times when you want the container to have access
      to your data. In these cases, you can link your local folders with
      folders inside the container. This particular container runs from the
      `/project` directory by default, so that's where we link our dashboard
      folder.
   3. the last 4 lines are arguments to the `create-predevals-data.R` script
      that lives inside the docker image. The arguments are:
      - `-h`: path to the hub (in the container)
      - `-c`: path to the config file (note that this is a relative path and it's
        coming from the dashboard folder)

>       - `-d`: the data file for the oracle output

Reviewer: ```suggestion
      - `-d`: the path to the data file or directory for the oracle output
```
? Or do you mean the destination directory? It's a little confusing because the source data is called oracle output.

Author: This is what I meant. And yes, I agree that it's confusing because the input is definitely called "output": https://docs.hubverse.io/en/latest/user-guide/target-data.html#oracle-output

```suggestion
      - `-d`: the path to the oracle data file
```

      - `-o`: the output directory

Once you have run this, you should see a folder similar to [the metrocast
dashboard predevals/data
branch](https://github.com/reichlab/metrocast-dashboard/tree/predevals/data).

(dashboard-local-site)=
### Generate site (docker)

The website is generated with [quarto](https://quarto.org), but instead of
providing a quarto template, we provide [a simplified
template](https://github.com/hubverse-org/hub-dashboard-template) that allows
hub administrators to add custom pages and styling while avoiding getting
bogged down in the intricacies of the quarto publishing system.

To build the website, we need three things:

> 1. The dashboard source markdown files and configuration file

Reviewer: might be good to clarify you mean the `site-config.yml`?

Author: ```suggestion
1. The dashboard source markdown files and `site-config.yml`
```

bare minimum steps required to build a dashboard website** independent of
operational GitHub workflows[^operations]. Once you understand these steps, then
you will be able to think about how and when to deploy them.

At the end of this exercise, you will have the source for a website that you
can explore interactively on your local machine. All of the data are embedded
within the same folder as the website, so you can explore the data structures
and configuration files that power the visualizations.

[^operations]: Specifically, this tutorial is explicitly about _the build
    process_, which is one part of the overall operations for the dashboards.
    We will build an interactive dashboard website, but in this context, we are
    not concerned of _when_ the site or data are built nor are we concerned of
    _deployment_. These will be covered in a later chapter on deployment with
    GitHub workflows.

## Rationale

Building a dashboard website requires three tools:

1. <https://github.com/hubverse-org/hub-dash-site-builder> builds the static website
1. <https://github.com/hubverse-org/hub-dashboard-predtimechart> builds the predtimechart data and options file

> 1. <https://github.com/hubverse-org/hubPredEvalsData-docker> builds the predevals data and options file

Reviewer 2: Are all of these lines supposed to start with "1"?

Author: I got frustrated rearranging the numbered list and this is a way to tell markdown to just number in order.

Using the hub and dashboard repositories you cloned in [the setup
step](#dashboard-local-setup), you can generate the data with these steps.

1. create evals output dir.
   ```bash
   mkdir -p $dash/data/evals
   ```
2. download the docker image (note that the image name needs to be in all lower case)
   ```bash
   docker pull --platform=linux/amd64 ghcr.io/hubverse-org/hubpredevalsdata-docker:main
   ```
3. Generate the evaluations.
   ```bash
   docker run --rm -it --platform=linux/amd64 \
     -v "$dash":"/project" \
     -v "$hub":"/hub" \
     ghcr.io/hubverse-org/hubpredevalsdata-docker:main \
     create-predevals-data.R \
       -h "/hub" \
       -c "predevals-config.yml" \
       -d "/hub/target-data/oracle-output.csv" \
       -o "data/evals"
   ```
   There is a lot going on in these lines, so let me break it down:
   1. `docker run --rm -it --platform=linux/amd64` tells docker to
      automatically remove the image when it's done, `-it` means that [it's an
      interactive TTY](https://stackoverflow.com/a/40026942), and
      `--platform=linux/amd64` allows it to run on newer macOS machines with
      ARM chips.

>    2. `-v` means "volume". It's pattern is `-v "/path/on/your/computer":"/path/in/image"`.

Reviewer 2: ```suggestion
   2. `-v` means "volume". Its pattern is `-v "/path/on/your/computer":"/path/in/image"`.
```
